experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsTypes.td

```diff
@@ -25,4 +26,36 @@ def Iterators_Stream : Iterators_Type<"Stream", "stream"> {
   let assemblyFormat = "`<` qualified($elementType) `>`";
 }
 
+//===----------------------------------------------------------------------===//
+// State of low-level iterators
+//===----------------------------------------------------------------------===//
+
+class Iterators_Iterator<string name, string typeMnemonic>
+    : Iterators_Type<name, typeMnemonic,
+                     [DeclareTypeInterfaceMethods<Iterators_IteratorInterface>]> {
+  let summary = "State of a Volcano-style iterator";
```

Comment:
Since we're playing games with visibility vs opacity boundaries, I'd go as far as spelling out the type in C++ and being explicit about what is opaque and visible at this level of abstraction.

Reply:
I am not sure what this means. What exactly are the visibility boundaries? What does "spelling out the type in C++" mean?

I am imagining things like this: eventually we will generate the implementation of `OpenOp` etc for the concrete type of its argument (which implement `IteratorInterface`). That implementation may in turn use `OpenOp`/ `NextOp`/`CloseOp` on the states of its uptream iterator(s), whose concrete types it knows (which implement `IteratorInterface`) but whose implementation is not generated yet.

Reply:
Offline discussion: this is meant as an addition in the documentation.

Reply:
Done with latest commit.